### PR TITLE
Frontlight progress bar

### PR DESCRIPTION
**Pull request by robert00s**
I prepared new frontlight widget. It should be easer to change frontlight in KO.
I'm not sure whether I put file frontlight.lua.

On generic devices:

On Kindle:

### CODE REVIEW

**Comment by AlanSP1** [comment]
Looks great. :+1: 

When working on frontlight, maybe it would be good to also change default values for touch zones for light toggling and status bar toggling? So they work by default for new users?

**Comment by robert00s** [comment]
When fronlight widget is open you can change the light level (immediately) by clicking on one of the black or white squares. So you can change light level e.g. from 10 to 20 by one click.
You can also click on max or min button to change light to max or min level.
It should work for all users with frontlight devices.
Sorry for my English.

**Comment by AlanSP1** [comment]
I'm sorry, I referenced @ycrichard PR https://github.com/koreader/koreader/pull/2284 i.e. this my comment here: https://github.com/koreader/koreader/pull/2284#issuecomment-261737635

It is out of scope of your PR, I probably should make new issue to solve this.

Your PR is great and big improvment. I would only retain possibility to enter numerical value like before, as light levels aren't always linear, at least not on my device. For example, there's big difference between values 19 and 20, but not between 18 and 19, as well as 20 and 21.

Maybe if we tap on number we could enter numerical value as before? Is it possible to make?

**Comment by Frenzie** [comment]
This looks fantastic visually, but I'm with @AlanSP1. Precise control is quite important. While the changes on my device (H2O) are gradual, anything over 20 and certainly anything over 30 can be classified as OMFG BRIGHT.[1] In the rare case that I enable the frontlight I'll probably use a value between 1 and 10 and precision is important. I prefer the KOReader system precisely because it offers more control than the vague wishy-washy slide business in Nickel.

I'm not entirely sure what the best solution would be GUI-wise. I'd propose a combination of a simple +/- would suffice, perhaps combined with the tapping on the number for manual input suggestion.

Below is one way to do it. The other would be to stick everything on one line, as in `Min`, `-`, `#`, `+`, `Max`.
frontlight

[1] Incidentally, the same applies to all backlit monitors I use. My Zenbook features the best laptop screen I've had since my first laptop from '04 because it can go sufficiently dark at night. @#$@# the brightness craze. :-P

**Comment by houqp** [comment]
Looks great!

@AlanSP1, with regards to  #2284, I am working on that one, so if @robert00s do that, it will conflict with my change :)

I also like the idea of adding +/- buttons. I think we can do this as a separate PR if you want.

File: frontend/apps/reader/modules/readerfrontlight.lua [hunk 1]
```diff
@@ -83,42 +82,10 @@ function ReaderFrontLight:onPanRelease(arg, ges)
...
-                    callback = function()
-                        self.fl_dialog:setInputText("")
-                        powerd:toggleFrontlight()
-                    end,
-                },
-                {
-                    text = _("Apply"),
-                    enabled = true,
-                    callback = function()
-                        self:fldialIntensity()
```
Comment (houqp) [inline frontend/apps/reader/modules/readerfrontlight.lua]: Is the fldialIntensity safe to delete?

**Comment by houqp** [comment]
Well done @robert00s , the only suggestion I have is to move frontlight.lua into `ui/widget` so it can be reused by other components like filemanager in the future ;)

**Comment by AlanSP1** [comment]
Great idea about having additional +/- buttons, but thinking about UI, it may be better to make them "+1" and "-1"? It may look more intuitive? What do you think?

@houqp good to know you're solving problem with overlapping tap zones.

**Comment by KenMaltby** [comment]
Why not make it a "long press" on the bar to bring up the numeric entry keyboard?  This would match the way "go to page" can function.

**Comment by houqp** [comment]
My only concern with long press feature is it will take at least 1 long press and 2 taps to get to a desired state, which is about the same amount of effort for tapping on the bar and use +/- to do minor adjustments.

**Comment by robert00s** [comment]
I prepared two version. With and without advanced button.
version 1 (with advanced button):

after click on advanced button:


version 2 (without advanced button)


In my opinion version 2 is much better and simpler.
Which version to implement?

**Comment by Frenzie** [comment]
Definitely version 2 because that's much clearer visually.

NB That just means I dislike the enormous "advanced" button; I still think that @AlanSP1's original suggestion of tapping the number could potentially be kind of cool. :-)

**Comment by AlanSP1** [comment]
@robert00s These looks really great.

Thinking about advanced button and seeing your images, I saw something potentially important for devices without hardware light toggling button (H20 is one). We can toggle light by tapping lower left zone (added here: https://github.com/koreader/koreader/pull/2284 ), but it is more logical that we also have light toggling ability through this interface also, i.e. that we still have old menu available which have software toggle button.

 So, I would keep a way to access advanced menu. Maybe not as first image shows, but through tapping numerical value. Also, numerical value can be a button itself which calls advanced menu. Number also can be within box, to stress that it is actually button which brings advanced menu.

In this way we still could have light toggling menu. It is a bit more clicks (one more than it is now), so maybe it's better to have dedicated toggle button? Maybe where is advanced button now? And maybe to make this button a bit smaller?

Just my thoughts...

**Comment by Frenzie** [comment]
@AlanSP1 Oh @#$@# you're right! I completely forgot about the fact that my most important reason for bringing up the frontlight dialog is to click toggle. I actually almost never enter a number.

**Comment by houqp** [comment]
Again, do we really need an advanced mode? I really don't think it would be more convenient than what @robert00s already implemented.

I would suggest move the number to the first row with +/- buttons. Then add a toggle button to middle the bottom row. We should definitely have toggle button 1 tap away since it's a heavily used feature.

**Comment by AlanSP1** [comment]
I'm fine without advanced mode, but light toggling is important. I mention old mode as advanced just as it already has light toggling function.

All in all, what @houqp suggest is very good interface in my opinion. Than we can pretty fast define light level and toggle light status.

**Comment by Frenzie** [comment]
Yup, sounds sensible to me.

Incidentally, we should also decide what happens if you press -1 when it's already at 1. The one alternative is that it turns off, the other is that nothing happens because you're already at the lowest light level and turning it on/off is a matter of the toggle.

**Comment by houqp** [comment]
> Incidentally, we should also decide what happens if you press -1 when it's already at 1. The one alternative is  that it turns off, the other is that nothing happens because you're already at the lowest light level and turning it on/off is a matter of the toggle.

I think it's better to enforce a minimal value with `-` button and use toggle to completely turn the light off. This avoids a tricky scenario where toggle button is pressed when light is already turned off through the `-` button. In this case, we don't know what value to restore to..

**Comment by robert00s** [comment]
Changes:
- frontlight without "advanced"
- added toggle button
- light level add to bottom line
- light level from min to max - when level is min then "-1" button is disabled, also +1 button is disabled when reach max level 
- frontlight.lua file move to ui/widget with new name frontlightwidget.lua

**Comment by Frenzie** [comment]
That looks very good indeed. Excellent!

**Comment by AlanSP1** [comment]
Looks great. :+1:

File: frontend/ui/widget/frontlightwidget.lua [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        radius = 0,
+        enabled = true,
+        width = self.screen_width * 0.20,
+        show_parent = self,
+        callback = function()
+            local powerd = Device:getPowerDevice()
+            powerd:toggleFrontlight()
+        end,
+    }
+    local empty_space = TextBoxWidget:new{
```
Comment (houqp) [inline frontend/ui/widget/frontlightwidget.lua]: Any reason not using the HorizontalSpan widget instead?

**Comment by houqp** [comment]
Merged, thanks for the awesome widget!

**Comment by houqp** [comment]
@robert00s I noticed that `prog_button` callbacks does not honor `self.fl_min` so I was able to set the text value to `0` even though it's caped to `1` by `fl_min`.

**Comment by mwoz123** [comment]
Will that work also for all Koreader devices (including Pocketbooks)?
Previously we had a brightness controll but only for Kindles and Kobos.

**Comment by houqp** [comment]
We still don't have code for controlling frontlight in pocketbook. We need someone with the device to help implement this.

**Comment by mwozny123** [comment]
I've PB and would be able to help if instructed what's TODO;)

**Comment by houqp** [comment]
If you have strace installed on your PB, try changing light settings after you attach to the native reader. The strace output should tell you what files or system calls it made to interface with the kernel. If you paste the strace output here, i can help analysis it.

**Comment by houqp** [comment]
You can pass the following arguments to strace: `strace -fitv -p PID_OF_NATIVE_READER`

**Comment by mwozny123** [comment]
list of PB processes:
```
  PID USER       VSZ STAT COMMAND
    1 root      2412 S    init
    2 root         0 SW   [kthreadd]
    3 root         0 SW   [ksoftirqd/0]
    5 root         0 SW   [kworker/u:0]
    6 root         0 SW   [rcu_kthread]
    7 root         0 SW<  [khelper]
    8 root         0 SW   [sync_supers]
    9 root         0 SW   [bdi-default]
   10 root         0 SW<  [kblockd]
   11 root         0 SW   [khubd]
   12 root         0 SW<  [cpufreq_uevent]
   13 root         0 SW<  [cfg80211]
   14 root         0 SW   [kworker/0:1]
   15 root         0 DW   [usb-hardware-sc]
   16 root         0 SW   [khungtaskd]
   17 root         0 SW   [kswapd0]
   18 root         0 SW   [fsnotify_mark]
   19 root         0 SW<  [crypto]
   32 root         0 SW<  [sun5i-spi.2]
   33 root         0 SW<  [f_mtp]
   34 root         0 SW   [file-storage]
   35 root         0 SW   [kinteractiveup]
   36 root         0 SW<  [kfantasy]
   37 root         0 SW   [kworker/u:1]
   38 root         0 SW   [kworker/0:2]
   39 root         0 SW   [mmcqd/1]
   40 root      2412 S    {rcS} /bin/sh /etc/init.d/rcS
   42 root         0 SW   [flush-179:0]
   51 root         0 DW   [epdc_update]
   56 root         0 SW   [irq/28-2-0024]
   62 root         0 SW<  [loop0]
   64 root      3392 S    ./pocketbook
   70 root      3388 S    ./pocketbook
   71 root     62620 S    ./pocketbook
   84 sreader  51444 S N  /ebrmain/bin/scanner.app
   86 sreader  63832 S <  /ebrmain/bin/bookshelf.app -i
  120 reader   91132 S N  /ebrmain/bin/universal_sync.app
  121 sreader  67124 S N  /ebrmain/bin/obreeysocial.app
  122 sreader  63300 S N  /ebrmain/bin/usage_stat.app
  123 reader   82968 S N  /ebrmain/bin/dropbox.app
  124 reader   43560 S N  /ebrmain/bin/control_panel_mgr.app
  127 sreader   111m S    /ebrmain/bin/explorer.app
  129 sreader  46124 S <  /ebrmain/bin/taskmgr.app
  130 root         0 SW   [RTW_CMD_THREAD]
  135 root      4828 S    /ebrmain/sbin/wpa_supplicant -ieth0 -Dnl80211,wext -
  136 root     41668 S    /ebrmain/bin/netagent net on
  186 reader    3820 S <  /mnt/ext1/applications/pb_sshd/usr/sbin/sshd -f /mnt
  243 reader    4028 S <  sshd: reader@notty
  246 reader    2412 S <  sh -c /mnt/ext1/applications/pb_sshd/usr/libexec/sft
  247 reader    3512 S <  {sftp-server} /mnt/ext1/applications/pb_sshd/usr/lib
  248 reader    3872 S <  sshd: reader@notty
  251 reader    2412 S <  sh -c /mnt/ext1/applications/pb_sshd/usr/libexec/sft
  252 reader    3512 S <  {sftp-server} /mnt/ext1/applications/pb_sshd/usr/lib
  285 reader    3872 S <  sshd: reader@notty
  288 reader    2416 S <  -sh
  294 reader    2416 R <  ps
```
The light is controlled via system menu. Not in system reader app.
I think it might be the process no. 124. Log in attachment:
[124.log.txt](https://github.com/koreader/koreader/files/674282/124.log.txt)

I used comand:

> strace -fitv -p 124 -o 124.log
> Process 124 attached with 3 threads - interrupt to quit
> Process 365 attached
> Process 124 suspended
> Process 366 attached
> Process 365 suspended
> Process 365 resumed
> Process 366 detached
> Process 124 resumed
> Process 365 detached
> Process 124 detached
> Process 307 detached
> Process 308 detached



For processes no. 60, 64, 70,71, 127 I had answer:
> attach: ptrace(PTRACE_ATTACH, ...): Operation not permitted

**Comment by houqp** [comment]
Nothing in control_panel_mgr seems related to frontlight control. Looks like you don't have enough permission to strace other processes. You will need a root shell to be able to strace arbitrary process.

**Comment by mwozny123** [comment]
Should I search for way to get a `root` on my PB?

Or anyway to change a light on PB, Koreader would have to be started with `root` privileges (which doesn't make sense)?

**Comment by houqp** [comment]
That's a good point, I assume koreader is started under reader user? I took a look at their inkview SDK header and can't find any API for frontlight control either: https://github.com/blchinezu/pocketbook-sdk/blob/master/PBSDK/include/inkview.h. Perhaps this functionality is not exposed to 3rd party applications :(

**Comment by mwozny123** [comment]
> I assume koreader is started under reader user? 

`whoami` returns `reader` in console. I assume koreader is started with the same privileges (from main menu).

Looking around the internet I found a ["pocketbook free SDK"](https://github.com/pocketbook-free/SDK_481) from user ["pocketbook"](https://github.com/pocketbookbut) not sure what that is.
It hasn't been updated over 1 year. 
And PB touch lux 3 (w626) firmware was recently [updated](http://www.pocketbook-int.com/us/support/pocketbook-touch-lux-3) to 5.14 version

**Comment by houqp** [comment]
Looks like it has a set frontlight message: https://github.com/pocketbook-free/SDK_481/blob/27b504517d44e3ab9ab3626f3de892cd24451da3/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h#L215

I vaguely remember pocketbook does RPC between different components using messages from the developer documentation. So this is probably the right way for 3rd party app to control the frontlight.

What firmware version are you using btw? Based on the readme, looks like there are 3 toolchains for PB: https://github.com/blchinezu/pocketbook-sdk/blob/master/README.md. Currently all koreader PB releases are built with SDK_481. This might be the cause of your random segfaults.

**Comment by houqp** [comment]
Now, we need to figure out how the RPC system works in pocketbook.

**Pull request merged**